### DISCUSSION TITLE
Ship Classification and Ship Icons

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#78] [Ideas]
This discussion is about two topics that relate to each other: What ship classifications should there be, and how should ship icons be handled?

### Ship Classifications

Ships are defined in plugins as classification templates, like a 'Galaxy class" or "Defiant class", to use Star Trek terminology. During a flight, these can be stamped out to create actual ships, like the USS Enterprise or USS Defiant. Most of the properties of the ship should be defined on the ship template itself, but I also want to have broad ranges of classifications to more easily make distinctions between different groups of ships.

These will be built-in to the codebase, and are not configurable by end-users (though we could add more classifications later on if deemed necessary)

Example classifications, based on ship size:
- Cruiser
- Frigate
- Scout
- Shuttle

I think this provides a good balance between not overwhelming with choices while giving a good range of options. I also intentionally avoided obvious military classifications, like "Destroyer" or "Battleship".

Any suggestions for other classifications or renaming the classifications I've chosen?

### Ship Icons

Ships will have icons that will show on the sensor grid, tactical maps, and other places. How should these icons be defined?

(Quick aside, since I know this will come up - I think Sensors will be tasked with identifying ships on the sensor grid, similar to how it is done in Empty Epsilon - they have to scan the ship to find out what its faction or classification is. Until that happens, it will use a built-in generic icon.)

We could just have each ship template define it's own icon. However, in my experience as a Flight Director, I most often want the ship's icon to correspond to its faction - Starfleet ships get a Starfleet logo, Klingon ships get a Klingon logo, etc.

Ship templates are not bound to one faction. There could be two opposing factions that happen to use the same classification of ships. Does that mean that the factions should decide what icon ships should have?

And should there be separate icons for each classification of ship, to make it easier to tell the classification from the sensor grid just by looking at the icon?

Here's my proposal: All of the above. Factions define icons for each level of classification (another reason to limit how many there are). Ship templates can define their own icon, which overrides the faction icon. 

Sound good?

**Comment by astrolamb-gaming** [comment]
Ship Classifications:
I don't really see your reasons for avoiding military classifications. If you want to do so, I suggest renaming "cruiser" to "explorer", and "frigate" to "escort", perhaps. I also don't totally understand the reasoning behind hardcoding the classifications. I do understand the desire to have general classifications that give a general sense of a ship's capabilities, but people will want to have more options that that. I expect they'll want battleships and dreadnaughts and carriers and who knows what else. 
I'll note that ship classifications IRL is not necessarily based on size (though there is a correlation), but on function. There's destroyers today that are bigger than some inter-war period cruisers, though if we're avoiding military designations I guess that's a moot point. 


Ship Icons:
Lots of ways to handle this. I _really_ like how Artemis takes the silhouette of a ship's 3d model as the icon (after it's been scanned), and uses color to indicate faction, though for players all non-player-faction ships are red in color. The GM screen has different colors for each faction, but the faction colors aren't great sometimes because artemis can have like 20+ sides, and the colors are just too similar sometimes.
One of my favorite military sci-fi books references colors being used for different factions, though there's only a like four factions in the sereis. How many factions will Nova support? If there's more than maybe six at a time, colors may not be the way to go.

Your proposal of an icon for each classification for each faction would work ok, but if ship templates can override, then how do you know what faction or class the ship is at a glance? Isn't that the point of having different icons?

Another point to consider: Should there be text associated with each ship icon? Artemis has the ship name written directly above the icon, which has pros and cons. I recommend some sort of toggle for whether or not text is displayed, and that the text should include the ship's name, classification, and faction. If text is displayed, and can be quickly toggled on/off, then I'm much more comfortable with the icon overrides, since a quick toggle and you can check specifics about that ship without having to exclusively rely on checking with science as to what is what. But I'm still more of a fan of the silhouette method for the icon.

**Comment by alexanderson1993** [comment]
> avoiding military classifications.

Not all ships are military, like freighters and passenger ships. I'm not opposed to using military-sounding classifications, though. It's just my preference.

> hardcoding the classifications

It depends on whether we use the classifications for configuring other things, like having factions include an icon for each classification. If the classification is just a name and isn't used elsewhere in the code or configuration, there's nothing wrong with allowing any value. 

> silhouette of a ship's 3d model as the icon

This is a good idea, and easy to automate. I still think there should be the option to override icons, both for factions and for ships. Sometimes the ship silhouette just doesn't look appealing, and a well-designed icon can be much easier to identify. Maybe it falls back on the silhouette if there are no override icons.

> all non-player-faction ships are red in color

I would prefer the player have proper IFF - Ships turn red if they are hostile, gray for disabled, green for friendly, yellow for... something. I also think there should be an option to choose the color scheme. So the Sensors officer can choose IFF, or they can choose to show the faction color, or they can remove colors altogether.

> How many factions will Nova support?

Unlimited, since plugins can add as many factions as they want. In [this blog post](https://thoriumsim.com/blog/thorium-novas-lore) I described 7 stereotypical factions that could be included in the main storyline.

> Isn't that the point of having different icons?

This is intended for special ships that are unique for a particular mission. For example, in one mission there's a standard freighter  ship called the "Outlaw Star" which the crew interacts with many times through the mission. Instead of using a generic ship symbol, that ship gets a special icon to differentiate it from other ships, for story purposes.


>  I recommend some sort of toggle for whether or not text is displayed

Agreed, both that there should be the option for text and that it can be turned off. Or have the ability to cycle through what the text says, like you suggested: name, classification, faction, and any other values we think might be valuable to see at a glance.

---

All great points. Thanks for the thoughtful analysis. Here's an amendment to my original proposal:

Ships can define any classification they like, and classifications on other ships will be provided as suggestions to make it more difficult to have a typo and accidentally create a new classification.

Factions can define a _single_ icon which is applied to all of their ships, but ships can still override with their own icon. And if there is no icon provided, the silhouette will be used automatically.

**Reply by astrolamb-gaming** [reply]
Personally I'd rather factions be associated with colors, and use icons to represent classifications/classes istead. The reverse just doesn't work as well, and if there's, say, a blue glow around all the icons in a particular fleet, it'll be obvious which faction they're from at a glance, while you'll still be able to determine what types of ships the fleet is composed of. 
To combat the issue of a color being used for multiple factions, the flight director could assign a color for each of the factions that will be encountered in a given mission. Or perhaps (or in addition) the captain/science officer could assign colors.

But I'd be ok with your amendmenet.

**Reply by alexanderson1993** [reply]
As a demonstration, here are two (poorly crafted) strategic maps, one where faction is indicated with color and another where faction is indicated with icons. Tell me which one you prefer.

![colorsf](https://user-images.githubusercontent.com/6558157/139340327-497c5001-ed98-4d34-ae03-3f64a9bbc29f.png)
![logos](https://user-images.githubusercontent.com/6558157/139340329-79b49172-a6ae-47c9-b39c-828081eadf78.png)

One other benefit of using icons to represent faction: you can use colors to represent other things, like IFF. And you could still have that blue glow to represent factions while still having the icons.

**Reply by alexanderson1993** [reply]
> icons to represent classifications/classes Instead.

One other problem with this - this brings us back to having to configure different kinds of ship classifications, since you need to define the icon associated with the ship class. I would... prefer to avoid that, especially for something so generic that applies to so many ships. It makes things really messy when you create a ship class in one plugin that a whole bunch of other plugins rely on.

**Comment by Xansta** [comment]
My preferences:
For tactical purposes, I'd like the screen to show ship-like icons so that one can see the direction they are pointing. So, on the helm and weapons tactical screen, you can, at a glance see what ships are pointing at at you, what ships are parallel, and what ships are pointed away from you. Also for tactical reasons, I prefer a simple color scheme to represent friends, enemies and neutrals. It might be nice to have a color (or something) representing armed vs. unarmed.

For strategic purposes, like on a longer range view, icons and/or colors can be used to show factions. Regions of space under the influence of a particular faction could be color coded for each faction. Regions might also show friendly, neutral, enemy with the possibility of gradients in between such as enemies with a cease fire agreement, neutrals negotiating to be friendly or neutrals considering joining an enemy faction or alignment.

Empty Epsilon (for the crew) starts ships as a generic gray/white triangle when they are not scanned - no color to indicate relationship, no shape to indicate ship (type or classification). Once the first scan is complete, the faction relationship and the ship type is visible (color for friend, enemy, neutral; pointy, rotatable icon for ship type). This scheme is used on any screen that shows ship icons (helm, weapons, science, relay). The GM screen has the same pointy/rotatable icons, but a different color is given to each faction. The second scan provides details useful in combat.

Instead of only two levels of scan detail, there could be a third level that uses the classification described. The first scan might show the ship classification (eg cruiser) and the friend/enemy/neutral status, the second scan provides the specific ship type (cruiser, interceptor class or cruiser, sniper class) along with the faction and the third scan provides details like shield frequencies, status of subsystems (damaged or healthy engines, etc).

I guess I'm trying to say that there can be more than one way to represent a ship depending on the role of a crew person and the level of detail available. Having these varying levels of detail to be represented also gives greater opportunity for multiple crew to assist in the scanning. Perhaps one is primarily focused on combat and the tactical situation, while another supports the overall strategic goals, thus needing a broader picture rather than some of the details.

Since this is a starship bridge simulation, knowledge of other ships may be limited due to the vastness of space. Current naval and air force systems have a large body or research behind them to help identify ships. In space, such databases may or may not be available on whatever the sensors are picking up. An initial abstract definition may classify a ship as a cruiser, but the second or third pass via sensors may refine that into a larger or smaller classification: what might seem like a large ship is reclassified once the sensors determine that the large surface area showing on sensors was really just a light sail and it's actually much smaller or a comparatively small ship may get reclassified due to the faction's penchant for using captured black holes as power sources.

As @astrolamb-gaming  mentioned, the classification of naval vessels relates more to purpose in the real world. I've seen some older naval vessel classifications that use tonnage or size to classify ships. Whatever we come up with may use different aspects, but I suggest we pick one primary characteristic as a way to classify ships. If we want to give a more alien flavor for story telling, perhaps we can have "our" method of classifying ships (say by size) and yet have to be aware of the friendly aliens way of classifying ships (by top speed) and the enemy aliens way of classifying ships (available destructive power). The crew may have to adapt if their sensors are repaired using one of these alien parts making them rethink their classification of ships.

**Reply by alexanderson1993** [reply]
### Icons

I love the _idea_ of seeing the direction of ships using icons, but it is _really_ hard to map 3D space to a 2D plane. How do you indicate that a ship is pointed up or down? How do you represent the orientation of the ship when dealing with pitch and roll, and not just yaw? 3D makes everything messier...

In Thorium Nova Prime, I addressed this by using direction-less icons which were replaced with 3D models when you zoom close enough. https://thoriumsim.com/blog/update-16-map-updates

It think that's a good enough solution, and we could add little lines or arrows to indicate direction as well. Heck, we could even 3D rotate a flat icon to indicate orientation, but from the right angle, the flat icon would disappear. Maybe a checkbox that lets you alternate between a flat view that only shows the yaw component of the ships orientation, and a 3D view that rotates the icon.

There's one more problem, though - this requires plugin creators to design icons that have a clear direction to them, and if they don't add a direction to their icon their logo looks silly being rotated around. Perhaps there could be another checkbox when configuring the ship icon that says "This icon indicates direction". When checked, the icon rotates. When not checked, the line/arrow appear.

Yeah, I think that's a viable solution. I'd love to hear your thoughts. 

### Classifications

I would prefer to not have different classifications for different factions. That adds a fair bit of complexity to the programming and configuration without adding a ton of value, as I see it. I could be convinced otherwise, but I think we can just hand-wave and say everyone in the universe classifies their ships by size.

### Sensor Scans

This should be discussed more in a future discussion, but I figure we can chat about it a bit here too. You bring up Empty Epsilon's scanning. I would love to do something totally different from that, based more on how Thorium Classic works.

In Thorium Classic, the sensors officer types text into a box with their scan query. "Nearby pirate ships." "Weapons on the USS Ranger." "Mutara Nebula Radiation." etc. The Flight Director or co-Flight Director types in an answer and sends it back.

Obviously this isn't feasible in a fully automated set up. Even doing raw text analysis would be complicated and tricky, where the computer can determine the intent of a scan query (but really awesome if we got it right). So what needs to happen is a combination of Thorium Classic and Empty Epsilon. There are preset scans which the sensors officer can pick from, with different options for scanning open space vs. scanning an object vs. scanning a specific kind of object, like a ship or planet.

So we could do something similar to what you described. The crew member selects a ship and scans for "Classification", which tells them how big the ship is. Then they have to scan for "Faction". These kinds of scans probably go pretty quick. Other scans that get more specific take longer to complete, like scanning for the max speed, weapons capabilities, or shield strength. All of these would have to be separate scans, but I imagine they could fire off multiple scans simultaneously. However, each scan drains their available power and having multiple concurrent scans slows down all of them - only so much sensors to go around, I suppose.

Once Sensors scans the ships, the properties that are identified are made available to the other stations when they look at the ships on their controls.

I would love thoughts on all of these. If you've got lots of thoughts on Sensors, it might make sense to just start another discussion too.

**Comment by Xansta** [comment]
### Icons
I think context will determine what's used. Some will get reused from context to context

**Huge scale map (multiple solar systems):**

- Star
- Planet
- Nebula
- Black Hole
- Unknown
- Any items that don't represent stellar phenomena could be represented by a faction icon representing knowledge in the database about the presence of that faction in the area rather than direct sensor scan results

**Large scale map (approximately solar system size):**

- Star
- Planet
- Planetoid
- Large asteroid
- Comet
- Unknown Energy Source 
  - Electromagnetic radiation 
  - Light
  - Other energy
  - Likely a ship, but it could be anything
- Again, known faction installations could be represented by a faction icon. At this range, friendly ships, stations, planets with orbiting sensors will all likely be transmitting data about themselves and their automated short range scanning information. In this friendly sensor bubble is where we might have non-directional classification by size ship icons (cruiser, frigate, scout, shuttle). Selecting these icons could give textual information available such as faction, ship type, direction of travel, bearing from us, speed of travel, relative speed to us, flight plan (maybe graphical, maybe just a listed destination). Regardless, the icon does not need to represent direction because at this range, the information has no relevance to any tactical decisions.

**Smaller than solar system, larger than automated short range sensors**
This is where we have to make some more challenging representative decisions. At this resolution of detail, knowing ship direction of the ships on the map at a glance may be critical to decision making. Knowing the faction and inferring the intent of a group of ships or individual ships may be more important. We could defer to the captain and make two views available: tactical and strategic (or call them something else, that's just what comes to mind). If we're pursuing the full 3D immersion, these views may need to have something other than "the bird's eye view" available to facilitate proper tactical deployment of the ship and/or friendly fleets in the area.

Short Range Sensors
The definition of short range may vary from ship to ship or perhaps from mission to mission, but the information portrayed in this map should definitely have some kind of representation of direction for ships. This does not have to be done via icon, but that makes the most sense to me. As discussed, the icon in a 3D context has drawbacks. Maybe in tactical situations, views will be reduced to 2D to allow for iconic values for ships with other visual conventions to represent the 3rd dimension. Maybe the 3rd dimension does not have to be explicitly displayed. For helm and for those operating various weapons and shields, it may be enough to know when we have an enemy in range, when we have an enemy targeted, when the enemy has us in range and/or when an enemy has us targeted. A projected weapon path (to or from) might be a nice visual cue. Shields may be available for energy focus front, back, port, starboard, dorsal, ventral, upper, lower, etc. and knowing where to activate/focus them in the event the shield automation fails (likely) would be good information for crew to have during a fight. 

Rendering in 3D for a viewscreen may suffice, but a tactical view is supposed to simplify the decisions necessary during combat. 

Some of these view concepts may reveal a basic difference in philosophy for how these bridge simulations should run. The starship Enterprise rarely blows up during the show. Without making everything a cakewalk, it will be a challenge to write automated missions with some sense of perilous adventure without running the risk of catastrophic mission failure. With a flight director, there can be some allowance for poor tactical or strategic decisions because the flight director can protect the crew from themselves or an inherently dangerous universe. If it's largely automated, it's easy for either the crew to destroy themselves or to take unforeseen actions that cause the mission to hang. When you have a group that has scheduled an outing to go on a simulated space adventure, you want them to succeed. That's hard to do if you grant them any significant degree of autonomy (I'm thinking the younger participants). Similarly, if a group of adults schedules a similar kind of outing, they'll get bored if there's minimal challenge (everything has some kind of safety factor protecting the ship and crew).

I tend to write missions where there's always a real possibility of getting destroyed or defeated or failing the mission with the idea that the journey itself is as entertaining as the success of the mission. This often means multiple attempts to complete a mission, learning from the deadly mistakes of the previous attempts. If it's written well, there should be enough clues given for the crew or crews to succeed

**Reply by alexanderson1993** [reply]
I think I agree with pretty much everything you describe here. Good tradeoffs, good alternatives and options. Giving options to the crew to let them decide how they want to view things makes the most sense to me.
